File: frontend/src/arkham/types/Question.ts
import * as JsonDecoder from 'ts.data.json';
import { Message, messageDecoder } from '@/arkham/types/Message';
import { FlavorText, flavorTextDecoder } from '@/arkham/types/FlavorText';

export type Question = ChooseOne | ChooseUpToN | ChooseSome | ChooseSome1 | ChooseN | ChooseOneAtATime | ChooseOneAtATimeWithAuto | ChooseDeck | ChooseUpgradeDeck | ChoosePaymentAmounts | ChooseAmounts | QuestionLabel | Read | PickSupplies | DropDown | PickScenarioSettings | PickCampaignSettings | ChooseOneFromEach;

export enum QuestionType {
  CHOOSE_ONE = 'ChooseOne',
  PLAYER_WINDOW_CHOOSE_ONE = 'PlayerWindowChooseOne',
  CHOOSE_ONE_FROM_EACH = 'ChooseOneFromEach',
  CHOOSE_UP_TO_N = 'ChooseUpToN',
  CHOOSE_SOME = 'ChooseSome',
  CHOOSE_SOME_1 = 'ChooseSome1',
  CHOOSE_N = 'ChooseN',
  CHOOSE_ONE_AT_A_TIME = 'ChooseOneAtATime',
  CHOOSE_ONE_AT_A_TIME_WITH_AUTO = 'ChooseOneAtATimeWithAuto',
  CHOOSE_UPGRADE_DECK = 'ChooseUpgradeDeck',
  CHOOSE_DECK = 'ChooseDeck',
  CHOOSE_PAYMENT_AMOUNTS = 'ChoosePaymentAmounts',
  CHOOSE_AMOUNTS = 'ChooseAmounts',
  QUESTION_LABEL = 'QuestionLabel',
  READ = 'Read',
  PICK_SUPPLIES = 'PickSupplies',
  DROP_DOWN = 'DropDown',
  PICK_SCENARIO_SETTINGS = 'PickScenarioSettings',
  PICK_CAMPAIGN_SETTINGS = 'PickCampaignSettings',
}

export type PickScenarioSettings = {
  tag: QuestionType.PICK_SCENARIO_SETTINGS;
}

export type PickCampaignSettings = {
  tag: QuestionType.PICK_CAMPAIGN_SETTINGS;
}

export type ChooseOne = {
  tag: QuestionType.CHOOSE_ONE;
  choices: Message[];
}

// The backend represents this as a nest list, but we flatten it and pass the flattened index
export type ChooseOneFromEach = {
  tag: QuestionType.CHOOSE_ONE_FROM_EACH;
  choices: Message[];
}

export type QuestionLabel = {
  tag: QuestionType.QUESTION_LABEL
  card: string | null
  label: string
  question: Question
}

export type Read = {
  tag: QuestionType.READ
  flavorText: FlavorText
  readChoices: ReadChoices
  readCards: string[] | null;
}

type Supply
  = 'Provisions'
  | 'Medicine'
  | 'Rope'
  | 'Blanket'
  | 'Canteen'
  | 'Torches'
  | 'Compass'
  | 'Map'
  | 'Binoculars'
  | 'Chalk'
  | 'Pendant'
  | 'Gasoline'
  | 'Pocketknife'
  | 'Pickaxe'

export const supplyDecoder = JsonDecoder.oneOf<Supply>([
  JsonDecoder.literal('Provisions'),
  JsonDecoder.literal('Medicine'),
  JsonDecoder.literal('Rope'),
  JsonDecoder.literal('Blanket'),
  JsonDecoder.literal('Canteen'),
  JsonDecoder.literal('Torches'),
  JsonDecoder.literal('Compass'),
  JsonDecoder.literal('Map'),
  JsonDecoder.literal('Binoculars'),
  JsonDecoder.literal('Chalk'),
  JsonDecoder.literal('Pendant'),
  JsonDecoder.literal('Gasoline'),
  JsonDecoder.literal('Pocketknife'),
  JsonDecoder.literal('Pickaxe')
], 'Supply')

export type PickSupplies = {
  tag: QuestionType.PICK_SUPPLIES
  pointsRemaining: number
  chosenSupplies: Supply[]
  choices: Message[]
}

export type DropDown = {
  tag: QuestionType.DROP_DOWN
  options: string[]
}

export type ChooseN = {
  tag: QuestionType.CHOOSE_N
  amount: number
  choices: Message[]
}

export type ChooseSome = {
  tag: QuestionType.CHOOSE_SOME
  choices: Message[]
}

export type ChooseSome1 = {
  tag: QuestionType.CHOOSE_SOME_1
  choices: Message[]
}

export type ChooseUpToN = {
  tag: QuestionType.CHOOSE_UP_TO_N
  amount: number
  choices: Message[]
}

export type ChooseOneAtATime = {
  tag: QuestionType.CHOOSE_ONE_AT_A_TIME
  choices: Message[]
}

export type ChooseOneAtATimeWithAuto = {
  tag: QuestionType.CHOOSE_ONE_AT_A_TIME_WITH_AUTO
  label: string
  choices: Message[]
}

export type ChoosePaymentAmounts = {
  tag: QuestionType.CHOOSE_PAYMENT_AMOUNTS
  label: string
  paymentAmountTargetValue: AmountTarget | null
  paymentAmountChoices: PaymentAmountChoice[]
}

export type AmountTarget
  = { tag: 'MaxAmountTarget', contents: number }
  | { tag: 'TotalAmountTarget', contents: number }
  | { tag: "MinAmountTarget", contents: number }
  | { tag: 'AmountOneOf', contents: number[] }

export type ChooseAmounts = {
  tag: QuestionType.CHOOSE_AMOUNTS
  label: string
  amountTargetValue: AmountTarget
  amountChoices: AmountChoice[]
}

export type ChooseUpgradeDeck = {
  tag: QuestionType.CHOOSE_UPGRADE_DECK
}

export type ChooseDeck = {
  tag: QuestionType.CHOOSE_DECK
}

export type AmountChoice = {
  choiceId: string
  label: string
  minBound: number
  maxBound: number
}

export const amountChoiceDecoder = JsonDecoder.object<AmountChoice>({
  choiceId: JsonDecoder.string(),
  label: JsonDecoder.string(),
  minBound: JsonDecoder.number(),
  maxBound: JsonDecoder.number(),
}, 'AmountChoice')

export const amountTargetDecoder = JsonDecoder.oneOf<AmountTarget>(
  [ JsonDecoder.object({ tag: JsonDecoder.literal('MaxAmountTarget'), contents: JsonDecoder.number()}, 'MaxAmountTarget')
  , JsonDecoder.object({ tag: JsonDecoder.literal('MinAmountTarget'), contents: JsonDecoder.number()}, 'MinAmountTarget')
  , JsonDecoder.object({ tag: JsonDecoder.literal('TotalAmountTarget'), contents: JsonDecoder.number()}, 'TotalAmountTarget')
  , JsonDecoder.object({ tag: JsonDecoder.literal('AmountOneOf'), contents: JsonDecoder.array(JsonDecoder.number(), 'number[]')}, 'AmountOneOf')
  ]
, 'AmountTarget')

export const chooseAmountsDecoder = JsonDecoder.object<ChooseAmounts>(
  {
    tag: JsonDecoder.literal(QuestionType.CHOOSE_AMOUNTS),
    label: JsonDecoder.string(),
    amountTargetValue: amountTargetDecoder,
    amountChoices: JsonDecoder.array(amountChoiceDecoder, 'AmountChoice[]'),
  }, 'ChooseAmounts',
)

export type PaymentAmountChoice = {
  choiceId: string
  investigatorId: string
  minBound: number
  maxBound: number
  title: string
}

export const paymentAmountChoiceDecoder = JsonDecoder.object<PaymentAmountChoice>({
  choiceId: JsonDecoder.string(),
  investigatorId: JsonDecoder.string(),
  minBound: JsonDecoder.number(),
  maxBound: JsonDecoder.number(),
  title: JsonDecoder.string(),
}, 'PaymentAmountChoice')

export const choosePaymentAmountsDecoder = JsonDecoder.object<ChoosePaymentAmounts>(
  {
    tag: JsonDecoder.literal(QuestionType.CHOOSE_PAYMENT_AMOUNTS),
    label: JsonDecoder.string(),
    paymentAmountTargetValue: JsonDecoder.nullable(amountTargetDecoder),
    paymentAmountChoices: JsonDecoder.array(paymentAmountChoiceDecoder, 'PaymentAmountChoice[]'),
  }, 'ChoosePaymentAmounts',
);

export const chooseUpgradeDeckDecoder = JsonDecoder.object<ChooseUpgradeDeck>(
  {
    tag: JsonDecoder.literal(QuestionType.CHOOSE_UPGRADE_DECK),
  },
  'ChooseUpgradeDeck',
);

export const chooseDeckDecoder = JsonDecoder.object<ChooseDeck>(
  {
    tag: JsonDecoder.literal(QuestionType.CHOOSE_DECK),
  },
  'ChooseDeck',
);

export const pickScenarioSettingsDecoder = JsonDecoder.object<PickScenarioSettings>(
  {
    tag: JsonDecoder.literal(QuestionType.PICK_SCENARIO_SETTINGS),
  },
  'PickScenarioSettings',
);

export const pickCampaignSettingsDecoder = JsonDecoder.object<PickCampaignSettings>(
  {
    tag: JsonDecoder.literal(QuestionType.PICK_CAMPAIGN_SETTINGS),
  },
  'PickCampaignSettings',
);

export const questionLabelDecoder: JsonDecoder.Decoder<QuestionLabel> = JsonDecoder.object<QuestionLabel>(
  {
    tag: JsonDecoder.literal(QuestionType.QUESTION_LABEL),
    label: JsonDecoder.string(),
    card: JsonDecoder.nullable(JsonDecoder.string()),
    question: JsonDecoder.lazy(() => questionDecoder)
  },
  'QuestionLabel',
);


export type ReadChoices
  = { tag: "BasicReadChoices", contents: Message[] }
  | { tag: "LeadInvestigatorMustDecide", contents: Message [] }


export const readChoicesDecoder: JsonDecoder.Decoder<ReadChoices> = JsonDecoder.oneOf<ReadChoices>( [
  JsonDecoder.object({
    tag: JsonDecoder.literal('BasicReadChoices'),
    contents: JsonDecoder.array(messageDecoder, 'Message[]')
  }, 'BasicReadChoices'),
  JsonDecoder.object({
    tag: JsonDecoder.literal('BasicReadChoicesN'),
    contents: JsonDecoder.tuple([JsonDecoder.number(), JsonDecoder.array(messageDecoder, 'Message[]')], 'contents')
  }, 'BasicReadChoicesN').map(({ contents }) => ({ tag: 'BasicReadChoices', contents: contents[1] })),
  JsonDecoder.object({
    tag: JsonDecoder.literal('BasicReadChoicesUpToN'),
    contents: JsonDecoder.tuple([JsonDecoder.number(), JsonDecoder.array(messageDecoder, 'Message[]')], 'contents')
  }, 'BasicReadChoicesUpToN').map(({ contents }) => ({ tag: 'BasicReadChoices', contents: contents[1] })),
  JsonDecoder.object({
    tag: JsonDecoder.literal('LeadInvestigatorMustDecide'),
    contents: JsonDecoder.array(messageDecoder, 'Message[]')
    }, 'LeadInvestigatorMustDecide')
], 'ReadChoices');

export const readDecoder: JsonDecoder.Decoder<Read> = JsonDecoder.object<Read>(
  {
    tag: JsonDecoder.literal(QuestionType.READ),
    flavorText: flavorTextDecoder,
    readChoices: readChoicesDecoder,
    readCards: JsonDecoder.nullable(JsonDecoder.array(JsonDecoder.string(), 'CardCodes[]'))
  },
  'Read',
);

export const pickSuppliesDecoder = JsonDecoder.object<PickSupplies>(
  {
    tag: JsonDecoder.literal(QuestionType.PICK_SUPPLIES),
    pointsRemaining: JsonDecoder.number(),
    chosenSupplies: JsonDecoder.array<Supply>(supplyDecoder, 'Supply[]'),
    choices: JsonDecoder.array<Message>(messageDecoder, 'Message[]'),
  },
  'PickSupplies',
);

export const dropDownDecoder = JsonDecoder.object<DropDown>(
  {
    tag: JsonDecoder.literal(QuestionType.DROP_DOWN),
    options: JsonDecoder.array(JsonDecoder.tuple([JsonDecoder.string(), JsonDecoder.succeed()], '[string, message]').map(([s,]) => s), 'string[]') //eslint-disable-line
  },
  'PickSupplies',
);

export const chooseOneDecoder = JsonDecoder.object<ChooseOne>(
  {
    tag: JsonDecoder.oneOf(
        [JsonDecoder.literal(QuestionType.CHOOSE_ONE)
        , JsonDecoder.literal(QuestionType.PLAYER_WINDOW_CHOOSE_ONE).map(() => QuestionType.CHOOSE_ONE)
        ], "ChooseOne.tag"),
    choices: JsonDecoder.array<Message>(messageDecoder, 'Message[]'),
  },
  'ChooseOne',
);

export const chooseOneFromEachDecoder = JsonDecoder.object<ChooseOneFromEach>(
  {
    tag: JsonDecoder.literal(QuestionType.CHOOSE_ONE_FROM_EACH),
    choices: JsonDecoder.array<Message[]>(JsonDecoder.array<Message>(messageDecoder, 'Message[]'), 'Message[][]').map(xs => xs.flat()),
  },
  'ChooseOneFromEach',
);

export const chooseSomeDecoder = JsonDecoder.object<ChooseSome>(
  {
    tag: JsonDecoder.literal(QuestionType.CHOOSE_SOME),
    choices: JsonDecoder.array<Message>(messageDecoder, 'Message[]'),
  },
  'ChooseSome',
);

export const chooseSome1Decoder = JsonDecoder.object<ChooseSome1>(
  {
    tag: JsonDecoder.literal(QuestionType.CHOOSE_SOME_1),
    choices: JsonDecoder.array<Message>(messageDecoder, 'Message[]'),
  },
  'ChooseSome1',
);

export const chooseNDecoder = JsonDecoder.object<ChooseN>(
  {
    tag: JsonDecoder.literal(QuestionType.CHOOSE_N),
    amount: JsonDecoder.number(),
    choices: JsonDecoder.array<Message>(messageDecoder, 'Message[]'),
  },
  'ChooseN',
);

export const chooseUpToNDecoder = JsonDecoder.object<ChooseUpToN>(
  {
    tag: JsonDecoder.literal(QuestionType.CHOOSE_UP_TO_N),
    amount: JsonDecoder.number(),
    choices: JsonDecoder.array<Message>(messageDecoder, 'Message[]'),
  },
  'ChooseUpToN',
);

export const chooseOneAtATimeDecoder = JsonDecoder.object<ChooseOneAtATime>(
  {
    tag: JsonDecoder.literal(QuestionType.CHOOSE_ONE_AT_A_TIME),
    choices: JsonDecoder.array<Message>(messageDecoder, 'Message[]'),
  },
  'ChooseOneAtATime',
);

export const chooseOneAtATimeWithAutoDecoder = JsonDecoder.object<ChooseOneAtATimeWithAuto>(
  {
    tag: JsonDecoder.literal(QuestionType.CHOOSE_ONE_AT_A_TIME_WITH_AUTO),
    label: JsonDecoder.string(),
    choices: JsonDecoder.array<Message>(messageDecoder, 'Message[]'),
  },
  'ChooseOneAtATimeWithAuto',
);

export const questionDecoder = JsonDecoder.oneOf<Question>(
  [
    chooseOneDecoder,
    chooseOneFromEachDecoder,
    chooseNDecoder,
    chooseSomeDecoder,
    chooseSome1Decoder,
    chooseUpToNDecoder,
    chooseOneAtATimeDecoder,
    chooseOneAtATimeWithAutoDecoder,
    chooseUpgradeDeckDecoder,
    chooseDeckDecoder,
    chooseAmountsDecoder,
    choosePaymentAmountsDecoder,
    questionLabelDecoder,
    readDecoder,
    pickSuppliesDecoder,
    dropDownDecoder,
    pickScenarioSettingsDecoder,
    pickCampaignSettingsDecoder,
    JsonDecoder.succeed().flatMap((f) => {
      return JsonDecoder.fail(f)
    })
  ],
  'Question',
);

File: backend/arkham-api/library/Arkham/Question.hs
{-# LANGUAGE DuplicateRecordFields #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE NoFieldSelectors #-}

module Arkham.Question where

import Arkham.Prelude hiding (maxBound, minBound)

import Arkham.Ability.Types
import Arkham.Campaigns.TheForgottenAge.Supply
import Arkham.Card
import Arkham.ChaosBagStepState
import Arkham.Id
import Arkham.SkillType
import Arkham.Source
import Arkham.Target
import Arkham.Tarot
import Arkham.Text
import Arkham.Window
import Control.Monad.Fail
import Data.Aeson.TH
import Data.UUID (nil)

data Component
  = InvestigatorComponent {investigatorId :: InvestigatorId, tokenType :: GameTokenType}
  | InvestigatorDeckComponent {investigatorId :: InvestigatorId}
  | AssetComponent {assetId :: AssetId, tokenType :: GameTokenType}
  deriving stock (Show, Eq, Ord, Data)

data GameTokenType = ResourceToken | ClueToken | DamageToken | HorrorToken | DoomToken
  deriving stock (Show, Eq, Ord, Data)

pattern ClueLabel :: InvestigatorId -> [msg] -> UI msg
pattern ClueLabel iid msgs <- ComponentLabel (InvestigatorComponent iid ClueToken) msgs
  where
    ClueLabel iid msgs = ComponentLabel (InvestigatorComponent iid ClueToken) msgs

pattern DamageLabel :: InvestigatorId -> [msg] -> UI msg
pattern DamageLabel iid msgs <- ComponentLabel (InvestigatorComponent iid DamageToken) msgs
  where
    DamageLabel iid msgs = ComponentLabel (InvestigatorComponent iid DamageToken) msgs

pattern AssetDamageLabel :: AssetId -> [msg] -> UI msg
pattern AssetDamageLabel aid msgs <- ComponentLabel (AssetComponent aid DamageToken) msgs
  where
    AssetDamageLabel aid msgs = ComponentLabel (AssetComponent aid DamageToken) msgs

pattern HorrorLabel :: InvestigatorId -> [msg] -> UI msg
pattern HorrorLabel iid msgs <- ComponentLabel (InvestigatorComponent iid HorrorToken) msgs
  where
    HorrorLabel iid msgs = ComponentLabel (InvestigatorComponent iid HorrorToken) msgs

pattern AssetHorrorLabel :: AssetId -> [msg] -> UI msg
pattern AssetHorrorLabel aid msgs <- ComponentLabel (AssetComponent aid HorrorToken) msgs
  where
    AssetHorrorLabel aid msgs = ComponentLabel (AssetComponent aid HorrorToken) msgs

pattern ResourceLabel :: InvestigatorId -> [msg] -> UI msg
pattern ResourceLabel iid msgs <- ComponentLabel (InvestigatorComponent iid ResourceToken) msgs
  where
    ResourceLabel iid msgs = ComponentLabel (InvestigatorComponent iid ResourceToken) msgs

data UI msg
  = Label {label :: Text, messages :: [msg]}
  | InvalidLabel { label :: Text }
  | TooltipLabel {label :: Text, tooltip :: Tooltip, messages :: [msg]}
  | CardLabel {cardCode :: CardCode, messages :: [msg]}
  | PortraitLabel {investigatorId :: InvestigatorId, messages :: [msg]}
  | TargetLabel {target :: Target, messages :: [msg]}
  | SkillLabel {skillType :: SkillType, messages :: [msg]}
  | SkillLabelWithLabel {label :: Text, skillType :: SkillType, messages :: [msg]}
  | EvadeLabel {enemyId :: EnemyId, messages :: [msg]}
  | FightLabel {enemyId :: EnemyId, messages :: [msg]}
  | EngageLabel {enemyId :: EnemyId, messages :: [msg]}
  | GridLabel {gridLabel :: Text, messages :: [msg]}
  | TarotLabel {tarotCard :: TarotCard, messages :: [msg]}
  | AbilityLabel
      { investigatorId :: InvestigatorId
      , ability :: Ability
      , windows :: [Window]
      , before :: [msg]
      , messages :: [msg]
      }
  | ComponentLabel {component :: Component, messages :: [msg]}
  | EndTurnButton {investigatorId :: InvestigatorId, messages :: [msg]}
  | StartSkillTestButton {investigatorId :: InvestigatorId}
  | SkillTestApplyResultsButton
  | ChaosTokenGroupChoice {source :: Source, investigatorId :: InvestigatorId, step :: ChaosBagStep}
  | EffectActionButton {tooltip :: Tooltip, effectId :: EffectId, messages :: [msg]}
  | Done {label :: Text}
  | SkipTriggersButton {investigatorId :: InvestigatorId}
  | CardPile {pile :: [PileCard], messages :: [msg]}
  deriving stock (Show, Eq, Data)

data PileCard = PileCard
  { cardId :: CardId
  , cardOwner :: Maybe InvestigatorId
  }
  deriving stock (Show, Eq, Data)

data PaymentAmountChoice msg = PaymentAmountChoice
  { choiceId :: UUID
  , investigatorId :: InvestigatorId
  , minBound :: Int
  , maxBound :: Int
  , title :: Text
  , message :: msg
  }
  deriving stock (Show, Eq, Data)

data AmountChoice = AmountChoice
  { choiceId :: UUID
  , label :: Text
  , minBound :: Int
  , maxBound :: Int
  }
  deriving stock (Show, Eq, Data)

data AmountTarget = MinAmountTarget Int | MaxAmountTarget Int | TotalAmountTarget Int | AmountOneOf [Int]
  deriving stock (Show, Eq, Data)

data Question msg
  = ChooseOne {choices :: [UI msg]}
  | PlayerWindowChooseOne {choices :: [UI msg]}
  | ChooseOneFromEach {groups :: [[UI msg]]}
  | ChooseN {amount :: Int, choices :: [UI msg]}
  | ChooseSome {choices :: [UI msg]}
  | ChooseSome1 {label :: Text, choices :: [UI msg]}
  | ChooseUpToN {amount :: Int, choices :: [UI msg]}
  | ChooseOneAtATime {choices :: [UI msg]}
  | ChooseOneAtATimeWithAuto {label :: Text, choices :: [UI msg]}
  | -- | Choosing payment amounts
    -- The core idea is that costs get broken up into unitary costs and we
    -- let the players decide how many times an individual player will pay
    -- the cost. The @Maybe Int@ is used to designate whether or not there
    -- is a target value. The tuple of ints are the min and max bound for
    -- the specific investigator
    ChoosePaymentAmounts
      { label :: Text
      , paymentAmountTargetValue :: Maybe AmountTarget
      , paymentAmountChoices :: [PaymentAmountChoice msg]
      }
  | ChooseAmounts
      { label :: Text
      , amountTargetValue :: AmountTarget
      , amountChoices :: [AmountChoice]
      , target :: Target
      }
  | ChooseUpgradeDeck
  | ChooseDeck
  | QuestionLabel {label :: Text, card :: Maybe CardCode, question :: Question msg}
  | Read {flavorText :: FlavorText, readChoices :: ReadChoices msg, readCards :: Maybe [CardCode]}
  | PickSupplies {pointsRemaining :: Int, chosenSupplies :: [Supply], choices :: [UI msg]}
  | DropDown {options :: [(Text, msg)]}
  | PickScenarioSettings
  | PickCampaignSettings
  deriving stock (Show, Eq, Data)

data ReadChoices msg
  = BasicReadChoices [UI msg]
  | BasicReadChoicesN Int [UI msg]
  | BasicReadChoicesUpToN Int [UI msg]
  | LeadInvestigatorMustDecide [UI msg]
  deriving stock (Show, Eq, Data)

data ChoosePlayerChoice = SetLeadInvestigator | SetTurnPlayer
  deriving stock (Show, Eq, Data)

evadeLabel
  :: (AsId enemy, IdOf enemy ~ EnemyId,  msg ~ Element (t msg), MonoFoldable (t msg))
  => enemy
  -> t msg
  -> UI msg
evadeLabel (asId -> enemy) (toList -> msgs) = EvadeLabel enemy msgs

fightLabel
  :: (AsId enemy, IdOf enemy ~ EnemyId,  msg ~ Element (t msg), MonoFoldable (t msg))
  => enemy
  -> t msg
  -> UI msg
fightLabel (asId -> enemy) (toList -> msgs) = FightLabel enemy msgs

targetLabel
  :: (Targetable target, msg ~ Element (t msg), MonoFoldable (t msg))
  => target
  -> t msg
  -> UI msg
targetLabel (toTarget -> target) (toList -> msgs) = TargetLabel target msgs

targetLabels
  :: (Targetable target, msg ~ Element (t msg), MonoFoldable (t msg))
  => [target]
  -> (target -> t msg)
  -> [UI msg]
targetLabels = flip mapTargetLabel

mapTargetLabel
  :: (Targetable target, msg ~ Element (t msg), MonoFoldable (t msg))
  => (target -> t msg)
  -> [target]
  -> [UI msg]
mapTargetLabel f = map (\c -> targetLabel c (f c))

mapTargetLabelWith :: Targetable target => (c -> target) -> (c -> [msg]) -> [c] -> [UI msg]
mapTargetLabelWith g f = map (uncurry targetLabel . (g &&& f))

concat
  [ deriveJSON defaultOptions ''GameTokenType
  , deriveJSON defaultOptions ''Component
  , deriveToJSON defaultOptions ''PaymentAmountChoice
  , [d|
      instance FromJSON msg => FromJSON (PaymentAmountChoice msg) where
        parseJSON = withObject "PaymentAmountChoice" $ \o -> do
          choiceId <- o .:? "choiceId" .!= nil
          investigatorId <- o .: "investigatorId"
          minBound <- o .: "minBound"
          maxBound <- o .: "maxBound"
          title <- o .: "title"
          message <- o .: "message"
          pure PaymentAmountChoice {..}
      |]
  , deriveJSON defaultOptions ''ChoosePlayerChoice
  , deriveToJSON defaultOptions ''AmountChoice
  , [d|
      instance FromJSON AmountChoice where
        parseJSON = withObject "AmountChoice" $ \o -> do
          choiceId <- o .:? "choiceId" .!= nil
          label <- o .: "label"
          minBound <- o .: "minBound"
          maxBound <- o .: "maxBound"
          pure AmountChoice {..}
      |]
  , deriveJSON defaultOptions ''AmountTarget
  , deriveJSON defaultOptions ''PileCard
  , deriveToJSON defaultOptions ''ReadChoices
  , [d|
      instance FromJSON msg => FromJSON (ReadChoices msg) where
        parseJSON (Array xs) = BasicReadChoices <$> parseJSON (Array xs)
        parseJSON (Object o) = $(mkParseJSON defaultOptions ''ReadChoices) (Object o)
        parseJSON other = fail $ "Unexpected json type: " <> show other
      |]
  , deriveJSON defaultOptions ''UI
  , deriveJSON defaultOptions ''Question
  ]
